api/firmware/attestation_internal_test.go

```diff
@@ -0,0 +1,206 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package firmware
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"testing"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/digitalbitbox/bitbox02-api-go/api/common"
+	"github.com/digitalbitbox/bitbox02-api-go/api/firmware/mocks"
+	"github.com/digitalbitbox/bitbox02-api-go/util/semver"
+	"github.com/stretchr/testify/require"
+)
+
+func unhex(s string) []byte {
+	b, err := hex.DecodeString(s)
+	if err != nil {
+		panic(err)
+	}
+	return b
+}
+
+// addAttestationPubkey temporarily modifies the attestation pubkeys list to add a mock entry for
+// testing.
+func addAttestationPubkey(identifier string, pubkeyInfo attestationPubkey) func() {
+	attestationPubkeys[identifier] = pubkeyInfo
+	return func() {
+		delete(attestationPubkeys, identifier)
+	}
+}
+
+func makeCertificate(rootPrivkey *btcec.PrivateKey, bootloaderHash []byte, devicePubkey []byte) []byte {
+	var certMsg bytes.Buffer
+	certMsg.Write(bootloaderHash)
+	certMsg.Write(devicePubkey)
+	sigHash := sha256.Sum256(certMsg.Bytes())
+	signature, err := rootPrivkey.Sign(sigHash[:])
+	if err != nil {
+		panic(err)
+	}
+	return append(signature.R.Bytes(), signature.S.Bytes()...)
+}
+
+func TestAttestation(t *testing.T) {
+
+	// Arbitrary values, they do not have any special meaning.
+	rootPubkeyIdentifier := unhex("78961bc16e9bc8144dd3db4fd8488e33ed93541f56943db17044e8f0a42b0ec1")
+	rootPrivateKey, rootPublicKey := btcec.PrivKeyFromBytes(
+		btcec.S256(),
+		unhex("15608dfed8e876bed1cf2599574ce853f7a2a017d19ba0aabd4bcba033a70880"),
+	)
+	bootloaderHash := unhex("3fdf2ff2dcbd31d161a525a88cb57641209c7eac2bc014564a03d34a825144f0")
+	devicePrivateKey, devicePublicKey := btcec.PrivKeyFromBytes(
+		elliptic.P256(),
+		unhex("9b1a4d293a6eef1960d8afab5e58dd581b135152ec3399bde9268fa23051321b"),
+	)
+	devicePubkeyBytes := devicePublicKey.SerializeUncompressed()[1:]
+
+	undo := addAttestationPubkey(
+		hex.EncodeToString(rootPubkeyIdentifier),
+		attestationPubkey{
+			pubkeyHex:                 hex.EncodeToString(rootPublicKey.SerializeUncompressed()),
+			acceptedBootloaderHashHex: hex.EncodeToString(bootloaderHash),
+		},
+	)
+	defer undo()
+
+	communication := &mocks.Communication{}
+	product := common.ProductBitBox02BTCOnly
+	device := NewDevice(
+		semver.NewSemVer(2, 0, 0),
+		&product,
+		&mocks.Config{}, communication, &mocks.Logger{},
+	)
+
+	// Query error.
+	expectedErr := errors.New("error")
+	communication.MockQuery = func([]byte) ([]byte, error) {
+		return nil, expectedErr
+	}
+	_, err := device.performAttestation()
+	require.Equal(t, expectedErr, err)
```

Comment:
I'd just check for non-nil error because if you later decide to wrap the actual return err with say fmt.Errorf("%w"), this will fail the test.

Reply:
On the other hand, if the function returns an error for a different reason, this unit test silently stops testing the intended branch. I'm okay with fixing the test in the future if it happens.